cluster/images/etcd/migrate/data_dir.go

```diff
@@ -0,0 +1,163 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+)
+
+// DataDirectory provides utilities for initializing and backing up an
+// etcd "data-dir" as well as managing a version.txt file to track the
+// etcd server version and storage verion of the etcd data in the
+// directory.
+type DataDirectory struct {
+	path        string
+	dir         *os.File
+	versionFile *VersionFile
+}
+
+// OpenOrCreateDataDirectory opens a data directory, creating the directory
+// if it doesn't not already exist.
+func OpenOrCreateDataDirectory(path string) (*DataDirectory, error) {
+	exists, err := exists(path)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		glog.Infof("data directory '%s' does not exist, creating it", path)
+		err := os.MkdirAll(path, 0755)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create data directory %s: %s", path, err)
+		}
+	}
+	dir, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open data directory %s: %s", path, err)
+	}
+
+	versionFile := &VersionFile{
+		path: filepath.Join(path, versionFilename),
+	}
+	return &DataDirectory{path, dir, versionFile}, nil
+}
+
+// Close releases the underlying File used by the DataDirectory.
+func (d *DataDirectory) Close() error {
+	return d.dir.Close()
+}
+
+// Initialize set the version.txt to the target version if the data
+// directory is empty. If the data directory is non-empty, no
+// version.txt file will be written since the actual version of etcd
+// used to create the data is unknown.
+func (d *DataDirectory) Initialize(target *EtcdVersionPair) error {
+	isEmpty, err := d.IsEmpty()
+	if err != nil {
+		return err
+	}
+	if isEmpty {
+		glog.Infof("data directory '%s' is empty, writing target version '%s' to version.txt", d.path, target)
+		err = d.versionFile.Write(target)
+		if err != nil {
+			return fmt.Errorf("failed to write version.txt to '%s': %s", d.path, err)
+		}
+		return nil
+	}
+	return nil
+}
+
+// Backup creates a backup copy of data directory.
+func (d *DataDirectory) Backup() error {
+	backupDir := fmt.Sprintf("%s.bak", d.path)
+	err := os.RemoveAll(backupDir)
+	if err != nil {
+		return err
+	}
+	err = os.MkdirAll(backupDir, 0755)
+	if err != nil {
+		return err
+	}
+	err = exec.Command("cp", "-r", d.path, backupDir).Run()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// IsEmpty returns true if the data directory is entirely empty.
+func (d *DataDirectory) IsEmpty() (bool, error) {
```

Comment:
`IsEmpty()` only works correctly on the first try, because each time it's called it exhausts another directory entry from `d.dir`.

Reply:
Thanks for catching @mml! I'll rewrite this to open the file only within functions where I need it opened.

Reply:
Actually, I think there's another level, too.  Pretty sure Readdirnames() will always return `.` and `..` even on an empty dir, except `/` has no `..`.

Reply:
I'll add tests. The references I was able to find to check if a directory is empty recommended this approach, but better to be safe and double check.

Reply:
I've added checks to data_dir_test to make sure this is correct. For posterity, here's where I found the approach https://stackoverflow.com/questions/30697324/how-to-check-if-directory-on-path-is-empty.  This is one cases where I'm a annoyed golang does not offer a convenience method.